pkg/services/ngalert/schedule/schedule.go

```diff
@@ -651,9 +639,27 @@ func (r *alertRuleRegistry) keyMap() map[models.AlertRuleKey]struct{} {
 }
 
 type alertRuleInfo struct {
-	evalCh  chan *evalContext
-	stopCh  chan struct{}
-	version int64
+	evalCh chan *evalContext
+	ctx    context.Context
+	Stop   context.CancelFunc
+}
+
+func newAlertRuleInfo(parent context.Context) *alertRuleInfo {
+	ctx, cancel := context.WithCancel(parent)
+	return &alertRuleInfo{evalCh: make(chan *evalContext), ctx: ctx, Stop: cancel}
+}
+
+// Eval signals the rule evaluation routine to perform the evaluation of the rule. Does nothing if the loop is stopped
+func (a *alertRuleInfo) Eval(t time.Time, version int64) bool {
+	select {
+	case a.evalCh <- &evalContext{
```

Comment:
This will panic when `a.evalCh` is closed, need to make sure that `Eval` does not run if `a.evalCh` is closed. I think it's best if `Eval` is responsible for closing `a.evalCh` then we don't need to think about writing into a closed channel, however, closing a channel should be done at most once so we will also need a marker.

Reply:
You're correct about panic, but we do not close the channel. This select makes the sender to stop sending data to a channel when the context is already cancelled and there is no consumer on other side. Due to nature of the scheduler this can happen when rule evaluation takes more time than it's evaluation interval which can cause evaluations to be piled up. When rule is deleted, we remove the rule from registry and stop the context. That causes those piled up evaluation to be cancelled immediatelly, and therefore, release the pointer to the channel so it could be collected by GC.